src/test/java/hudson/plugins/git/GitSCMTest.java

```diff
@@ -1637,6 +1637,46 @@ public void testSha1NotificationBranches() throws Exception {
         notifyAndCheckBranch(project, commit1, branchName, 1, git);
     }
 
+
+    @Issue("JENKINS-17348")
+    public void testPurgeDeletedBranch() throws Exception {
+        final FreeStyleProject project = setupProject("*/*", false);
+        final GitSCM scm = (GitSCM) project.getScm();
+        scm.getExtensions().add(new PruneStaleBranch());
+
+        final String commitFile1 = "commitFile1";
+        commit(commitFile1, johnDoe, "Commit number 1");
+        assertTrue("scm polling should detect commit 1",
+                project.poll(listener).hasChanges());
+        build(project, Result.SUCCESS, commitFile1);
+
+        git.checkout().ref("master").branch("topic-1").execute();
+        final String commitFile2 = "commitFile2";
+        commit(commitFile2, johnDoe, "Commit number 2 on topic-1");
+        assertTrue("scm polling should detect commit 2",
+                project.poll(listener).hasChanges());
+        FreeStyleBuild b2 = build(project, Result.SUCCESS, commitFile2);
+        BuildData data = b2.getAction(BuildData.class);
+        assertTrue("topic-1 branch should have been added", data.buildsByBranchName.containsKey("origin/topic-1"));
+        assertTrue("master branch should have been kept", data.buildsByBranchName.containsKey("origin/master"));
+
+        git.checkout().ref("master").execute();
+        git.deleteBranch("topic-1");
+
+        final String commitFile3 = "commitFile3";
+        commit(commitFile3, johnDoe, "Commit number 3");
+        assertTrue("scm polling should detect commit 3",
+                project.poll(listener).hasChanges());
+        FreeStyleBuild b3 = build(project, Result.SUCCESS, commitFile3);
+        data = b3.getAction(BuildData.class);
+        assertTrue("topic-1 branch should have been purged", !data.buildsByBranchName.containsKey("origin/topic-1"));
```

Comment:
`assertFalse` might be more idiomatic. (Or you can go the Hamcrest route.)

Would it make more sense to just assert something about the textual contents of `build.xml`, so as to avoid forcing test references to `BuildData`? IIUC the problem is not about some incorrect behavior, it is just about the build metadata steadily growing.